File: javascript/src/flyweight/flyweight.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { ConcreteFlyweight } from "./concrete-flyweight.js";
import { FlyweightFactory } from "./flyweight-factory.js";

describe("Flyweight", () => {
  it("the same key yields the same object", () => {
    const factory = new FlyweightFactory();
    assert.equal(factory.getFlyweight("a"), factory.getFlyweight("a"));
    assert.notEqual(factory.getFlyweight("a"), factory.getFlyweight("b"));
  });

  it("the pool grows only with distinct keys", () => {
    const factory = new FlyweightFactory();
    for (const key of ["x", "y", "x", "x", "z"]) {
      factory.getFlyweight(key);
    }
    assert.equal(factory.count(), 3);
  });

  it("operation combines intrinsic and extrinsic state", () => {
    assert.equal(new ConcreteFlyweight("q").operation(42), "ConcreteFlyweight(q) with extrinsic state 42");
  });
});

File: java/src/test/java/com/penapereira/patterns/flyweight/FlyweightTest.java
package com.penapereira.patterns.flyweight;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotSame;
import static org.junit.jupiter.api.Assertions.assertSame;

import org.junit.jupiter.api.Test;

class FlyweightTest {

    @Test
    void theSameKeyYieldsTheSameObject() {
        FlyweightFactory factory = new FlyweightFactory();
        assertSame(factory.getFlyweight("a"), factory.getFlyweight("a"));
        assertNotSame(factory.getFlyweight("a"), factory.getFlyweight("b"));
    }

    @Test
    void thePoolGrowsOnlyWithDistinctKeys() {
        FlyweightFactory factory = new FlyweightFactory();
        for (String key : new String[] {"x", "y", "x", "x", "z"}) {
            factory.getFlyweight(key);
        }
        assertEquals(3, factory.count());
    }

    @Test
    void operationCombinesIntrinsicAndExtrinsicState() {
        assertEquals("ConcreteFlyweight(q) with extrinsic state 42", new ConcreteFlyweight("q").operation(42));
    }
}
